### PR TITLE
Add "isn't" operator alongside the 'is' operator

### PR DESCRIPTION
**Pull request by harlanhaskins**
This patch adds a brand new operator that's going to revolutionize Swift programming: `isn't`. It works exactly like the `is` expression, only in reverse!

```swift
if x isn't Int {
  print("Good! I'm getting tired of Ints anyway")
}
```

Of course, if x is provably an `Int`, then we'll see the diagnostic:

```
warning: 'isn't' test is always false.
if x isn't Int {
   ~ ^     ~~~  
```

### CODE REVIEW

**Comment by segiddins** [comment]
Thank you for opening this pull request. The Swift user group has determined that it is out of scope for Swift 4.0. Please re-open the pull request once we start planning for the next major release of Swift.

**Comment by bitjammer** [comment]
@swift-ci Please smoke test

**Comment by bitjammer** [comment]
I think we'll need a blog post on this absolute game-changer. I would suggest the title, "You may not like it, but this 'isn't' what peak performance looks like"

File: lib/Parse/Lexer.cpp [hunk 1]
```diff
@@ -613,7 +613,20 @@ void Lexer::lexIdentifier() {
...
-  tok Kind = kindOfIdentifier(StringRef(TokStart, CurPtr-TokStart), InSILMode);
+  auto Str = StringRef(TokStart, CurPtr-TokStart);
+  tok Kind = kindOfIdentifier(Str, InSILMode);
+
+  // We don't want to support "isnt" without an apostrophe
+  if (Kind == tok::kw_isnt) {
+    Kind = tok::identifier;
+  }
+
+  if (Kind == tok::identifier && Str == "isn" &&
```
Comment (CodaFi) [inline lib/Parse/Lexer.cpp]: This creates a lexing ambiguity with the planned `ain't` keyword, and the `whomst'd've'ly'yaint'nt'ed'ies's'y'es` feature.

**Comment by AnthonyMDev** [comment]
I don't know how Swift is even usable without this.

**Comment by patchthecode** [comment]
Any problem in making it `!is` ?

**Comment by harlanhaskins** [comment]
The name `isn't` is paramount to this feature.

**Comment by CodaFi** [comment]
I agree.  I've searched Swift Evolution and, judging by word usage alone, `isn't` is one of the most talked-about features *ever*.  I don't even think this needs to go through evolution, I think it can just be merged as-is.

**Comment by AnthonyMDev** [comment]
Well you're at it, can we add a `might be` operator?

**Comment by harlanhaskins** [comment]
We'd also have to add `mightn't be`

**Comment by bitjammer** [comment]
Re: `mightn't be`, if you're going to pull in floating point semantics, I recommend consulting with @stephentyrone.

**Comment by dduan** [comment]
Nit: "isn't" has the same problem with "is". They only work for 3rd person singular subjects. This keyword should be `ain't`.

**Comment by bitjammer** [comment]
I don't see any machinery here to handle patterns and collection types, for which I would expect to see `aren't`, since the subject would more naturally be "they". You can address this in a follow up if you prefer.

**Comment by ahyattdev** [comment]
Swift is literally unusable without the `isn't'st'd've` operator.

**Comment by harlanhaskins** [comment]
Thanks for the fun, y'all.

**Comment by Pre1** [comment]
a possible extension for `is`, `isn't` operator:

```
if x [is | isn't] arrived { // async operation

} 
```

**Comment by capnslipp** [comment]
I don't know if this is the right time or place to bring this up, and my apologies if I'm stepping on any toes here, but it's surprised me again and again that Swift doesn't have an `asif` operator— _you know, a conditional evaluation operate that decides which codepath to take based on the likelihood of a language feature existing.

File: test/expr/cast/as_coerce.swift [hunk 2]
```diff
@@ -89,6 +89,19 @@ Double(1) as Double as String // expected-error{{cannot convert value of type 'D
...
+if 5 isn't String { // expected-warning{{'isn't' test always succeeds}}
+}
+
+if 5 isn't Int { // expected-warning{{'isn't' test always fails}}
+}
+
+struct Right {}
+let thatBoy = Right()
+
+if thatBoy ain't Right { // expected-warning{{'isn't' test always fails}}
```
Comment (macbellingrath) [inline test/expr/cast/as_coerce.swift]: 💯 :1st_place_medal:

**Pull request closed without merging**